Core/clim-basic/ports.lisp

```diff
@@ -171,107 +181,184 @@
   (with-slots (properties) port
     (setf (getf properties indicator) value)))
 
+;;; This function determines the sheet to which the pointer event
+;;; should be delivered. The right thing is not obvious:
+;;;
+;;; - we may assume that event-sheet is set correctly by port
+;;; - we may find the innermost sheet's child and deliver to it
+;;; - we may deliver event to sheet's graft and let it dispatch
+;;;
+;;; Third option would require a default handle-event method to call
+;;; handle-event on its child under the cursor. For now we implement
+;;; the second option with the innermost child. In general case we
+;;; need z-ordering for both strategies. -- jd 2019-08-21
+(defun compute-pointer-event-sheet (event &aux (sheet (event-sheet event)))
+  (labels ((gpes-rec (sheet x y)
+             (if-let ((child (child-containing-position sheet x y)))
+               (multiple-value-call #'gpes-rec
+                 child (untransform-position (sheet-transformation child) x y))
+               sheet)))
+    (get-pointer-position (sheet event)
+      (gpes-rec (event-sheet event) x y))))
+
+;;; Function is responsible for making a copy of an immutable event
+;;; and adjusting its coordinates to be in the target-sheet
+;;; coordinates. Optionally it may change event's class.
+(defun dispatch-event-copy (target-sheet event &optional new-class
+                            &aux (sheet (event-sheet event)))
+  (if (eql target-sheet sheet)
+      (dispatch-event sheet event)
+      (let ((new-event (shallow-copy-object event)))
+        (get-pointer-position (target-sheet new-event)
+          (setf (slot-value new-event 'x) x
+                (slot-value new-event 'y) y
+                (slot-value new-event 'sheet) target-sheet))
+        (when new-class
+          (change-class new-event new-class))
+        (dispatch-event target-sheet new-event))))
+
+;;; Boundary-events generated by port apply only to mirrored
+;;; sheets. For sheets without a mirror we must synthetize
+;;; boundary-events. Here is an outline what it does:
+;;;
+;;; pointer-exit-event first exits all unmirrored sheets from the last
+;;; port-pointer-sheet upwards and finally it dispatches the event for
+;;; the mirror. port-pointer-sheet is NIL after this operation.
+;;;
+;;; port-enter-event dispatches the event for the mirror and then to
+;;; each parent of the innermost sheet under the pointer from the
+;;; highest unmirrored ancestor. port-pointer-sheet is assigned to the
+;;; innermost child.
+;;;
+;;; Other pointer events result in finding a common ancestor of the
+;;; old port-pointer-sheet and the innermost sheet under the
+;;; pointer. First we synthetize exit events and then enter events.
```

Comment:
after the thought, I want to test synthetize-boundary-events a little more and if possible simplify the code.

Reply:
OK, I think this one should be fine now.